central/telemetry/marketing/gatherer.go

```diff
@@ -0,0 +1,137 @@
+package marketing
+
+import (
+	"context"
+	"time"
+
+	apDataStore "github.com/stackrox/rox/central/authprovider/datastore"
+	groupDataStore "github.com/stackrox/rox/central/group/datastore"
+	roles "github.com/stackrox/rox/central/role/datastore"
+	si "github.com/stackrox/rox/central/signatureintegration/datastore"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+	mpkg "github.com/stackrox/rox/pkg/telemetry/marketing"
+	"github.com/stackrox/rox/pkg/version"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+type marketing struct {
+	telemeter mpkg.Telemeter
+	period    time.Duration
+	ticker    *time.Ticker
+	stopSig   concurrency.Signal
+	ctx       context.Context
+	cancel    context.CancelFunc
+	userAgent string
+}
+
+var (
+	m    *marketing
+	once sync.Once
+)
+
+// InitGatherer initializes the periodic telemetry data gatherer.
+func InitGatherer(t mpkg.Telemeter, p time.Duration) {
+	once.Do(func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		m = &marketing{
+			telemeter: t,
+			period:    p,
+			userAgent: "central/" + version.GetMainVersion(),
+			ctx:       sac.WithAllAccess(ctx),
+			cancel:    cancel,
+			stopSig:   concurrency.NewSignal(),
+		}
+	})
+}
+
+// Singleton returns the previously initialized telemeter instance.
+func Singleton() *marketing {
+	return m
+}
+
+func (m *marketing) loop() {
+	for !m.stopSig.IsDone() {
```

Comment:
Probably a go expert should take a look at this, but is this concurrency safe? To my understanding accessing a struct from inside a go routine is not safe, and value passing between different go routines should be done via channels.

Reply:
The values are passed by channels in the methods of this primitive from the concurrency package.
Would be nice to have a look from a go expert, indeed.

Reply:
What I mean is that you are accessing the struct (e.g. `m.ticket` `m.telemeter`) from concurrent go routines. I think this is a potential data race, which would require either a mutex of some sort.

Reply:
`m.ticker` is set once before starting the loop in the method `Start()` below, then, its internal channel is being read.
`m.telemeter` is set once at the initialization and then used to call `Enqueue()` from the Segment API, which writes a message to a channel. Doesn't look like needing a mutex.